Add tenable spec

ndowmon opened the pull request:
FWIW we are currently in the process of making a big conversion in this integration anyway, but it's a good opportunity to review the data model here.

Here's the J1 Vuln data model: https://support.jupiterone.io/hc/en-us/articles/360041429733-Data-Model-for-Vulnerability-Management

**Current**:
![Screen Shot 2021-08-24 at 10 03 51 AM](https://user-images.githubusercontent.com/15333061/130632807-d64437d7-45a8-4f20-8bf6-02531628451d.png)


**Proposed**:
![Screen Shot 2021-08-24 at 10 32 03 AM](https://user-images.githubusercontent.com/15333061/130636205-7cdc2234-9b4d-4e7a-88d4-624ce2e3f7b8.png)

ndowmon commented on docs/spec/src/containers/index.ts:
```diff
@@ -0,0 +1,94 @@
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const executionHandler = () => {};
+
+export const containerSpec: Step<IntegrationStepExecutionContext>[] = [
```
I've left containers as-is here. But please feel free to comment here.

ndowmon commented on docs/spec/src/vulnerabilities/index.ts:
```diff
@@ -0,0 +1,136 @@
+     * ENDPOINT: n/a
+     * PATTERN: Build Child Relationships
+     */
+    id: 'build-asset-vuln-relationships',
```
This step is not very streamlined - creates 1 direct relationship and 2 mapped relationships. I wonder if mapping the target `host` back to _two_ entities in this integration is an indication of an anti-pattern.

ndowmon commented on docs/spec/src/vulnerabilities/index.ts:
```diff
@@ -0,0 +1,136 @@
+      {
+        _type: 'tenable_asset_is_host',
+        sourceType: 'tenable_asset',
+        _class: RelationshipClass.IS,
```
Per the data model, `HostAgent -PROTECTS-> Host`, even though I've put `IS` here. 

This one is sort of weird, because this `tenable_asset` is not actually a `HostAgent` (although it can be _discovered_ by host agents). Rather, it's a representation of a target asset. It's almost as though its properties should be added to the target `Host`, if we can find it. But that definitely opens up a lot more complexity in this data model which I'm not sure provides a lot of value implemented.

Thoughts on `PROTECTS` vs `IS`? And, whether `tenable_asset` should actually be a `HostAgent` or something different?

aiwilliams replied:
The query that we need to support, [similar for Qualys](https://jupiterone.atlassian.net/browse/INT-1405): `FIND Host THAT !PROTECTS HostAgent`. I think it needs to be `PROTECTS`. In the case where there is no actual agent installed on the host, it seems we need some entity to represent the "asset" (really that's just a record of tenable's knowledge that a host exists). I think this is something we need to talk through.

ndowmon commented:
Updated the vuln data model:

![Screen Shot 2021-08-31 at 4 29 01 PM](https://user-images.githubusercontent.com/15333061/131585971-b406fb52-e866-495e-8655-748889254a21.png)


Also added `toImplementSpec()` matcher.

ceelias commented:
That test addition is a solid add!

I think this is consistent with what we discussed. 

Not sure it makes a big difference here but we did discuss earlier about whether or not the `Agent` should be both a `HostAgent` and a `Scanner`. Based off what you currently have I don't think it should also be a `Scanner`, since it doesn't have any `SCAN` relationships. That makes me think that the `Agent` also shouldn't have a `IDENTIFIED` and that should just be left for the `Scanner`. Again def not make or break but wanted to bring this piece back up.

ndowmon commented:
>That makes me think that the `Agent` shouldn't have the `IDENTIFIED` relationship and that should just be left for the `Scanner`. What're your thoughts here?

Seems OK to me. There's definitely some weirdness happening with scanners & agents... we'll have to discuss when it comes to implementing them.

>Another thing I want to bring back up is the discussion of what the `Asset` should be. I think for now we should go with `Record` but I think we should consider creating an `Asset` class since we would like to enforce some consistency between these types.

In this case, there won't be consistency between this integration and other vuln scanners, so the `_class` consistency is not as critical. I assume (read: hope) that users will be filtering based on `_type` here to be more precise (in most cases, I think it makes the most sense to filter based on `_type`).

>What do you think the `Host` mapping should map to for the filter? Do you think we should include `Device` as target as well? These things don't need to be decided upon right now, but just wanted to throw that out there.

When we get to building these mapped relationships, we are going to need to build a number of different filters. We'll cross that bridge when we get there 😄
